docs/architecture/psa-migration/psa-limitations.md

```diff
@@ -0,0 +1,224 @@
+
+The RSA/PK API is in principle more flexible than the PSA Crypto API. The
+following sub-sections study whether and how this matters in practice.
+
+### Use in X.509
+
+RFC 4055 Section 3.1 defines the encoding of RSA-PSS that's used in X.509.
+It allows independently specifying the message hash (also used for encoding
+hash), the MGF (and its hash if MGF1 is used), and the salt length (plus an
+extra parameter "trailer field" that doesn't vary in practice"). These can be
+encoded as part of the key, and of the signature. If both encoding are
+presents, all values must match except possibly for the salt length, where the
+value from the signature parameters is used.
+
+In Mbed TLS, RSA-PSS parameters can be parsed and displayed for various
+objects (certificates, CRLs, CSRs). During parsing, the following properties
+are enforced:
+- (the extra "trailer field" parameter must has its default value)
+- the mask generation function is MGF1
+- encoding hash = message hashing algorithm (may differ from MGF1 hash)
+
+When it comes to cryptographic operations, only two things are supported:
+- verifying the signature on a certificate from its parent;
+- verifying the signature on a CRL from the issuing CA.
+
+The verification is done using `mbedtls_pk_verify_ext()`.
+
+Note: since X.509 parsing ensures that message hash = encoding hash, and
+`mbedtls_pk_verify_ext()` use encoding hash = mgf1 hash, it looks like all
+three hash algorithms must be equal, which would be good news as it would
+match a limitation of the PSA API. (TODO: double-check that.)
+
+Also, since we only use signature verification, the fact that PSA accepts any
+valid salt length means that no valid certificate would be wrongly rejected;
+however it means that signatures that don't match the announced salt length
+would be incorrectly accepted. At first glance, it looks like this doesn't
+allow an attacker to forge certificates, so this might be acceptable in
+practice, while not fully implementing all the checks in the standard. (TODO:
+triple-check that.)
+
+It is unclear what parameters people use in practice.
+
+### Use in TLS
+
+In TLS 1.2 (or lower), RSA-PSS signatures are never used, except via X.509.
+
+In TLS 1.3, RSA-PSS signatures can be used directly in the protocol (in
+addition to indirect use via X.509). It has two sets of three signature
+algorithm identifiers (for SHA-256, SHA-384 and SHA-512), depending of what
+the OID of the public key is (rsaEncryption or RSASSA-PSS).
+
+In both cases, it specifies that:
+- the mask generation function is MGF1
+- all three hashes are equal
+- the length of the salt MUST be equal to the length of the digest algorithm
+
+When signing, the salt length picked by PSA is the one required by TLS 1.3
+(unless the key is unreasonably small).
+
+When verifying signatures, again is doesn't look like accepting any salt
+length would give an attacker any advantage, but this must be triple-checked
+(TODO).
```

Comment:
I am not aware of any attack on RSA that would allow an attacker to craft a signature with a shorter salt (e.g. empty salt) but not a signature with a larger salt. Nonetheless, https://github.com/ARMmbed/mbedtls/pull/5010 fixes PSS verification in PSA to require the standard salt length as documented (and adds a variant that accepts an arbitrary salt length).